Add explanations (#32)
canonical/ubuntu-packaging-guide: docs/explanation/upstream-and-downstream.rst

Upstream & Downstream
=====================

An :term:`Ubuntu` installation consists of :term:`packages <Package>` - copied and unpacked onto
the target machine. The :term:`Ubuntu` project - packages, distributes and maintains software of
thousands of :term:`open-source <Open Source Software>` projects for users, ready to install.
The collection of :term:`Ubuntu` :term:`packages <Package>` is derived from the collection of
:term:`packages <Package>` maintained by the community-driven :term:`Debian` project.

An important duty of a :term:`Ubuntu` :term:`Package` :term:`Maintainer` is to collaborate with
the :term:`open-source <Open Source Software>` projects the :term:`Ubuntu`:term:`packages <Package>`
are derive from. Especially with :term:`Debian` by keeping the :term:`Ubuntu` copies of
:term:`packages <Package>` up-to-date and by sharing improvements made in :term:`Ubuntu` back up
to :term:`Debian`.

Terminology
-----------

Ubuntu delta
~~~~~~~~~~~~

Ubuntu delta (noun):
    A modification to an :term:`Ubuntu` :term:`Package` that is derived from a :term:`Debian`

>     :term:`Package`.

Reviewer: From what I understand (I fully expect I'm wrong), the delta is not so much a modification to an Ubuntu package, but rather the difference (in the mathematical sense) between the Debian package and the Ubuntu one - which might represent one modification, or several.

Author: That is totally correct :) Can I change something to make a reader more confident after reading the Ubuntu delta explanation?

----

In the context of :term:`Open Source Software` development, the terminology of a stream that
carries modifications, improvements, and code is used. It describes the relationship and direction
of changes made between projects. This stream originates (upwards) from the original project
(and related entities like :term:`Source Code`, authors, and :term:`maintainers <Maintainer>`) and

> flows downwards to projects (and associated entities) that depend on it.

Reviewer: This feels like a bit like a preamble, so I'd suggest putting it between the Terminology heading and the delta subheading

Author: Sounds good to me. It was originally there, before I pushed it, but I thought it is weird that I first explain this analogy and then introduce "Ubuntu delta".

Author: I put this to resolved as you already moved it yourself.

    
    *Examples*
        - *"We upstreamed the patch."*
        - *"Can you upstream the bugfix?"*

.. _Downstream:

Downstream
~~~~~~~~~~

*downstream (noun)*
    Similar to :ref:`upstream_noun`: A software project(s) (and associated entities) that depend on
    another software project directly or indirectly.

    *Example*
        - :term:`Ubuntu` is a downstream of :term:`Debian` and there are many downstreams
          of :term:`Ubuntu`.

    *Usage note*
        - The :ref:`adjective/adverb form <downstream_adjective_adverb>` is much more commonly used.
        - There can be many layers. For example, *Kubuntu* is a :term:`flavour <Ubuntu flavours>`
          of :term:`Ubuntu`, therefore *Kubuntu* and :term:`Ubuntu` are downstreams of
          :term:`Debian`.

.. _downstream_adjective_adverb:

*downstream (adjective, adverb)*
    Similar to :ref:`upstream_adjective_adverb`: Something (usually a code modification like 
    a :term:`Patch`) that flows in the direction or is relative to a software project farther
    away from the original software project.

    *Examples*
        - :term:`Ubuntu` is a downstream project of :term:`Debian`.

>         - *"The bug is already patched downstream."*
>         - *"The bug was reported by a downstream user."*
>         - *"Downstream maintainers have submitted a bugfix."*
>         - *"The change may affect downstream users."*

Reviewer: ```suggestion
        - The bug is already patched downstream.
        - The bug was reported by a downstream user.
        - Downstream maintainers have submitted a bugfix.
        - The change may affect downstream users.
```

> *downstream (verb)*

Reviewer: ```suggestion
Downstream (verb):
```

    Similar to :ref:`upstream_verb`: Sending something (usually a :term:`Patch`) downstream
    that originated from an upstream project.

>     *Example*

Reviewer: ```suggestion
    *Example*:
```

>         - "We downstreamed the :term:`Patch`."

Reviewer: ```suggestion
        - We downstreamed the :term:`Patch`.
```

> Why upstream changes?

Reviewer: ```suggestion
Why do we upstream changes?
```

---------------------

.. note::

>     The following list does not aim for completeness. There are certainly more good arguments 

Reviewer: ```suggestion
    The following list does not aim for completeness. There are plenty more good arguments 
```

Reviewer: ("certainly more good" could be read like there are better arguments that aren't listed)

    for why changes should be upstreamed.

> - **Decreased Maintenance Complexity**: Think of any :term:`Ubuntu` :term:`Package` derived from a 

Reviewer: ```suggestion
- **Decreased maintenance complexity**: Think of any :term:`Ubuntu` :term:`Package` derived from a 
```

  :term:`Debian` :term:`Package` that carries a :term:`delta <Ubuntu Delta>`. Every time the
  :term:`Debian` :term:`Package` gets updated, the :term:`Ubuntu` :term:`Package` may be subject
  to a :term:`Merge Conflict` when the changes to the :term:`Debian` :term:`Package` get applied to

>   the :term:`Ubuntu` :term:`Package`. Upstreaming changes reduces the maintenance cost to resolve

Reviewer: ```suggestion
  the :term:`Ubuntu` :term:`Package`. By upstreaming changes we reduce the maintenance cost to resolve
```

  :term:`Merge Conflicts <Merge Conflict>` when they occur.

> - **Quality Assurance & Security**: Any changes that get upstreamed will also be subject to the

Reviewer: ```suggestion
- **Quality assurance and security**: Any changes that get upstreamed will also be subject to the
```

  quality assurance of the upstream project and the testing coverage that the user base of the
  upstream project provides. This increases the likelihood of discovering regressions/bugs/unwanted
  behaviour (especially security-related bugs). Also, be aware that an unpatched
  :term:`security vulnerability <Common Vulnerabilities and Exposures>` in any system could lead to
  the indirect exposure of other systems.

> - **Common Benefit**: By syncing the :term:`Debian` :term:`packages <Package>` into the

Reviewer: ```suggestion
- **Common benefit**: By syncing the :term:`Debian` :term:`packages <Package>` into the
```
